karapace/client.py

```diff
@@ -0,0 +1,179 @@
+"""
+karapace - utils
+
+Copyright (c) 2022 Aiven Ltd
+See LICENSE for details
+"""
+from aiohttp import ClientSession
+from collections.abc import Mapping
+from typing import Awaitable, Callable, Optional, Union
+from urllib.parse import urljoin
+
+import logging
+import ssl
+
+log = logging.getLogger(__name__)
+Path = str
+Headers = dict
+JsonData = object  # Type of the result after parsing JSON
+
+
+async def _get_aiohttp_client() -> ClientSession:
+    return ClientSession()
+
+
+class Result:
+    def __init__(
+        self,
+        status: int,
+        json_result: JsonData,
+        headers: Optional[Mapping] = None,
+    ) -> None:
+        # We create both status and status_code so people can be agnostic on whichever to use
+        self.status_code = status
+        self.status = status
+        self.json_result = json_result
+        self.headers = headers if headers else {}
+
+    def json(self) -> JsonData:
+        return self.json_result
+
+    def __repr__(self) -> str:
+        return "Result(status=%d, json_result=%r)" % (self.status_code, self.json_result)
+
+    @property
+    def ok(self) -> bool:
+        return 200 <= self.status_code < 300
+
+
+class Client:
+    def __init__(
+        self,
+        server_uri: Optional[str] = None,
+        client_factory: Callable[[], Awaitable[ClientSession]] = _get_aiohttp_client,
+        server_ca: Optional[str] = None,
+    ) -> None:
+        self.server_uri = server_uri or ""
+        # aiohttp requires to be in the same async loop when creating its client and when using it.
+        # Since karapace Client object is initialized before creating the async context, (in
+        # kafka_rest_api main, when KafkaRest is created), we can't create the aiohttp here.
+        # Instead we wait for the first query in async context and lazy-initialize aiohttp client.
+        self.client_factory = client_factory
+
+        self.ssl_mode: Union[None, bool, ssl.SSLContext]
+        if server_ca is None:
+            self.ssl_mode = False
+        else:
+            self.ssl_mode = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+            self.ssl_mode.load_verify_locations(cafile=server_ca)
+        self._client: Optional[ClientSession] = None
+
+    def path_for(self, path: Path) -> str:
```

Comment:
mypy was inferring that `path_for` was a `Callable[[str], AnyStr]`, so I moved it to a method to get typing working.